File: day_4/part_1.py
import numpy as np


FILENAME = 'input.txt'

def fileread(filename: str, debug: bool = False) -> np.array:
    with open(filename, 'r') as f:
        array = np.array([[1 if char == '@' else 0 for char in line.strip()] 
                      for line in f], dtype=int)
    if debug == True:
        print(f"Array shape: {array.shape}")  # Debug line
        print(f"Array type: {type(array)}")    # Debug line
    return array

def gridsearch(grid: np.array, debug: bool = False) -> int:
    accessible = 0

    directions = [(di, dj) for di in [-1, 0, 1] for dj in [-1, 0, 1] if not (di == 0 and dj == 0)]

    for i in range(len(grid)):
        for j in range(len(grid[0])):
            if debug == True: 
                print(f"Searching cells: {i, j}")
        
            if grid[i][j] == 1:
                adjacent_paper_count = 0
            
                for di, dj in directions:
                    ni, nj = i + di, j + dj
                    if debug == True:
                        print(f"Searching neighbouring cells: {ni, nj}")
                    if 0 <= ni < len(grid) and 0 <= nj < len(grid[0]):
                        adjacent_paper_count += grid[ni][nj]
            
                if adjacent_paper_count < 4:
                    accessible += 1
    
    if debug == True:
        print(f"Accessible cells: {accessible}")
        print(f"Directions to search: {directions}")
    return accessible

def main():
    array = fileread(FILENAME)
    result = gridsearch(array)
    print(f"Accessible paper rolls: {result}")
    return result

if __name__ == '__main__':    
    main()



    

File: day_4/part_1.jl
function read_grid(filename)
  lines = readlines(filename)
  nrows = length(lines)
  ncols = length(lines[1])
  matrix = zeros(Int, nrows, ncols)
  for (i, line) in enumerate(lines)
    for (j, char) in enumerate(line)
      matrix[i,j] = (char == '@') ? 1 : 0
    end 
  end 
  return matrix
end

function grid_search(grid)  # No colon in Julia function definitions
  accessible = 0
  directions = [(di, dj) for di in -1:1 for dj in -1:1 if !(di == 0 && dj == 0)]
  
  nrows = size(grid, 1)
  ncols = size(grid, 2)
  
  for i in 1:nrows  # Julia is 1-indexed, use 1:n not range()
    for j in 1:ncols
      if grid[i, j] == 1
        adjacent_paper_count = 0
        
        for (di, dj) in directions  # No colon after 'in directions'
          ni, nj = i + di, j + dj
          # Check bounds (1-indexed!)
          if 1 <= ni <= nrows && 1 <= nj <= ncols
            adjacent_paper_count += grid[ni, nj]
          end
        end
        
        if adjacent_paper_count < 4
          accessible += 1
        end
      end
    end
  end
  
  return accessible
end

grid = read_grid("input.txt")
result = grid_search(grid)
println(result)
